lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -5919,50 +5919,60 @@ LogicalResult LayerBlockOp::verify() {
   }
 
   // Verify the body of the region.
-  Block *body = getBody(0);
-  bool failed = false;
-  body->walk<mlir::WalkOrder::PreOrder>([&](Operation *op) {
-    // Skip nested layer blocks.  Those will be verified separately.
-    if (isa<LayerBlockOp>(op))
-      return WalkResult::skip();
-    // Check all the operands of each op to make sure that only legal things are
-    // captured.
-    for (auto operand : op->getOperands()) {
-      // Any value captured from the current layer block is fine.
-      if (operand.getParentBlock() == body)
-        continue;
-      // Capture of a non-base type, e.g., reference, is allowed.
-      FIRRTLBaseType baseType = dyn_cast<FIRRTLBaseType>(operand.getType());
-      if (!baseType)
-        return WalkResult::advance();
-      // Capturing a non-passive type is illegal.
-      if (!baseType.isPassive()) {
-        auto diag = emitOpError()
-                    << "captures an operand which is not a passive type";
-        diag.attachNote(operand.getLoc()) << "operand is defined here";
-        diag.attachNote(op->getLoc()) << "operand is used here";
-        failed = true;
-        return WalkResult::advance();
-      }
-    }
-    // Ensure that the layer block does not drive any sinks.
-    if (auto connect = dyn_cast<FConnectLike>(op)) {
-      auto dest = getFieldRefFromValue(connect.getDest()).getValue();
-      if (dest.getParentBlock() == body)
+  auto result = getBody(0)->walk<mlir::WalkOrder::PreOrder>(
+      [&](Operation *op) -> WalkResult {
+        // Skip nested layer blocks.  Those will be verified separately.
+        if (isa<LayerBlockOp>(op))
+          return WalkResult::skip();
+
+        // Check all the operands of each op to make sure that only legal things
+        // are captured.
+        for (auto operand : op->getOperands()) {
+          // Any value captured from the current layer block is fine.
+          if (auto *definingOp = operand.getDefiningOp())
+            if (getOperation()->isAncestor(definingOp))
+              continue;
+
+          // Capture of a non-base type, e.g., reference, is allowed.
+          FIRRTLBaseType type = dyn_cast<FIRRTLBaseType>(operand.getType());
+          if (!type)
+            continue;
+
+          // Capturing a non-passive type is illegal.
+          if (!type.isPassive()) {
+            auto diag = emitOpError()
+                        << "captures an operand which is not a passive type";
+            diag.attachNote(operand.getLoc()) << "operand is defined here";
+            diag.attachNote(op->getLoc()) << "operand is used here";
+            return WalkResult::interrupt();
+          }
+        }
+
+        // Ensure that the layer block does not drive any sinks outside.
+        if (auto connect = dyn_cast<FConnectLike>(op)) {
+          // ref.define is allowed to drive probes outside the layerblock.
```

Comment:
> And I think due to lexical scoping, the destination MUST reside "above" -- you can't define into a child layer but we don't need to check that because that can't happen. Forgive my slowly thinking this through 👍 .

Yep!

> I think propassign is banned not directly under a module currently, so probably nothing to worry about there (?).

Yes that's true, although actually I think propassign should be valid under a layerblock. I should be able to do something like this:
```
extmodule Foo:
  input p : String
 
module Top:
  layerblock A:
    inst foo of Foo
    propassign foo.p, String("...")
```

We need to ban the capture and escape of properties until we figure out how to lower these things.

I might allow propassign under layerblock in a followup PR.